迁移聊天记录数据的脚本占用内存与数据量相关，数据量大时候会直接out of memory

假设50w条聊天记录。

> 试试这个分支？https://github.com/noneplugin/nonebot-plugin-chatrecorder/tree/fix/memory
> 
> 你也可以自己调整一下每次迁移多少数据。

这个分支的内存占用是降低了很多，但是最后报错了。
```
08-27 13:16:28 [WARNING] 902a51ac4032_add_session_py | 正在迁移聊天记录数据，请不要关闭程序...
08-27 13:18:59 [WARNING] 902a51ac4032_add_session_py | 聊天记录数据迁移完成！
08-27 13:18:59 [ERROR] uvicorn | Traceback (most recent call last):
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 552, in _prepare_and_execute
    self._rows = await prepared_stmt.fetch(*parameters)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/asyncpg/prepared_stmt.py", line 176, in fetch
    data = await self.__bind_execute(args, 0, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/asyncpg/prepared_stmt.py", line 241, in __bind_execute
    data, status, _ = await self.__do_execute(
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/asyncpg/prepared_stmt.py", line 230, in __do_execute
    return await executor(protocol)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "asyncpg/protocol/protocol.pyx", line 201, in bind_execute
asyncpg.exceptions.NotNullViolationError: column "session_id" of relation "nonebot_plugin_chatrecorder_messagerecord" contains null values

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 585, in execute
    self._adapt_connection.await_(
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 125, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 185, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 564, in _prepare_and_execute
    self._handle_exception(error)
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 515, in _handle_exception
    self._adapt_connection._handle_exception(error)
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 802, in _handle_exception
    raise translated_error from error
sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.IntegrityError: <class 'asyncpg.exceptions.NotNullViolationError'>: column "session_id" of relation "nonebot_plugin_chatrecorder_messagerecord" contains null values

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/migangbot/.venv/lib/python3.11/site-packages/starlette/routing.py", line 677, in lifespan
    async with self.lifespan_context(app) as maybe_state:
  File "/usr/local/lib/python3.11/contextlib.py", line 204, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot/drivers/fastapi.py", line 166, in _lifespan_manager
    await self._lifespan.startup()
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot/drivers/_lifespan.py", line 36, in startup
    await self._run_lifespan_func(self._startup_funcs)
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot/drivers/_lifespan.py", line 30, in _run_lifespan_func
    await cast(ASYNC_LIFESPAN_FUNC, func)()
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot_plugin_datastore/db.py", line 104, in init_db
    await upgrade(config, "head")
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot_plugin_datastore/script/command.py", line 245, in upgrade
    await run_migration()
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot_plugin_datastore/script/utils.py", line 123, in run_migration
    await connection.run_sync(do_run_migrations, plugin_name)
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/engine.py", line 884, in run_sync
    return await greenlet_spawn(fn, self._proxied, *arg, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 192, in greenlet_spawn
    result = context.switch(value)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot_plugin_datastore/script/utils.py", line 115, in do_run_migrations
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/migangbot/.venv/lib/python3.11/site-packages/alembic/runtime/environment.py", line 922, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/migangbot/.venv/lib/python3.11/site-packages/alembic/runtime/migration.py", line 624, in run_migrations
    step.migration_fn(**kw)
  File "/migangbot/.venv/lib/python3.11/site-packages/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py", line 133, in upgrade
    with op.batch_alter_table(
  File "/usr/local/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/migangbot/.venv/lib/python3.11/site-packages/alembic/operations/base.py", line 375, in batch_alter_table
    impl.flush()
  File "/migangbot/.venv/lib/python3.11/site-packages/alembic/operations/batch.py", line 113, in flush
    fn(*arg, **kw)
  File "/migangbot/.venv/lib/python3.11/site-packages/alembic/ddl/postgresql.py", line 184, in alter_column
    super().alter_column(
  File "/migangbot/.venv/lib/python3.11/site-packages/alembic/ddl/impl.py", line 229, in alter_column
    self._exec(
  File "/migangbot/.venv/lib/python3.11/site-packages/alembic/ddl/impl.py", line 193, in _exec
    return conn.execute(  # type: ignore[call-overload]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1412, in execute
    return meth(
           ^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/sql/ddl.py", line 181, in _execute_on_connection
    return connection._execute_ddl(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1524, in _execute_ddl
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1844, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1984, in _exec_single_context
    self._handle_dbapi_exception(
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2339, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 585, in execute
    self._adapt_connection.await_(
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 125, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 185, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 564, in _prepare_and_execute
    self._handle_exception(error)
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 515, in _handle_exception
    self._adapt_connection._handle_exception(error)
  File "/migangbot/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 802, in _handle_exception
    raise translated_error from error
sqlalchemy.exc.IntegrityError: (sqlalchemy.dialects.postgresql.asyncpg.IntegrityError) <class 'asyncpg.exceptions.NotNullViolationError'>: column "session_id" of relation "nonebot_plugin_chatrecorder_messagerecord" contains null values
[SQL: ALTER TABLE nonebot_plugin_chatrecorder_messagerecord ALTER COLUMN session_id SET NOT NULL]
(Background on this error at: https://sqlalche.me/e/20/gkpj)

```

--- a/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
+++ b/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
@@ -28,10 +28,22 @@ def upgrade() -> None:
     with Session(op.get_bind()) as session:
         logger.warning("正在迁移聊天记录数据，请不要关闭程序...")
 
+        session_key_id_map = {}
+        # 读取已经存在的 session
+        for session_obj in session.scalars(select(SessionModel)).all():
+            session_key = (
+                session_obj.bot_id,
+                session_obj.bot_type,
+                session_obj.platform,
+                session_obj.level,
+                session_obj.id1,
+                session_obj.id2,
+                session_obj.id3,
+            )
+            session_key_id_map[session_key] = session_obj.id
+
         # 每次迁移的数据量为 10000 条
         migration_limit = 10000
-
-        session_key_id_map = {}
         count = session.query(MessageRecord).count()
         logger.info(f"聊天记录数据总数：{count}")
         for i in range(count // migration_limit + 1):
@@ -72,7 +84,7 @@ def upgrade() -> None:
                 session_key = (bot_id, bot_type, platform, level, id1, id2, id3)
                 # 保存 message id 和 session key 的对应关系
                 message_id_session_key_map[message[8]] = session_key
-                if session_key not in bulk_insert_sessions:
+                if session_key not in session_key_id_map and session_key not in bulk_insert_sessions: # 去重
                     bulk_insert_sessions[session_key] = {
                         "bot_id": bot_id,
                         "bot_type": bot_type,
@@ -84,28 +96,10 @@ def upgrade() -> None:
                     }
 
             if bulk_insert_sessions:
-                # 读取已经存在的 session
-                for session_obj in session.scalars(select(SessionModel)).all():
-                    session_key = (
-                        session_obj.bot_id,
-                        session_obj.bot_type,
-                        session_obj.platform,
-                        session_obj.level,
-                        session_obj.id1,
-                        session_obj.id2,
-                        session_obj.id3,
-                    )
-                    session_key_id_map[session_key] = session_obj.id
+                # 插入新的 session
+                session.execute(insert(SessionModel), list(bulk_insert_sessions.values()))
 
-                # 更新新插入的 session
-                new_sessions_dict = [
-                    session_dict
-                    for key, session_dict in bulk_insert_sessions.items()
-                    if key not in session_key_id_map  # 去重
-                ]
-                if new_sessions_dict:
-                    session.execute(insert(SessionModel), new_sessions_dict)
-
+                # 更新已经存在的 session
                 for session_obj in session.scalars(select(SessionModel)).all():
                     session_key = (
                         session_obj.bot_id,

--- a/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
+++ b/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
@@ -28,87 +28,106 @@ def upgrade() -> None:
     with Session(op.get_bind()) as session:
         logger.warning("正在迁移聊天记录数据，请不要关闭程序...")
 
-        statement = select(MessageRecord)
-        messages = session.scalars(statement).all()
-
-        bulk_insert_sessions = {}
-        message_id_session_key_map = {}
-        for message in messages:
-            bot_id = message.bot_id
-            bot_type = message.bot_type
-            platform = message.platform
-            level = "LEVEL0"
-            if message.detail_type == "private":
-                level = "LEVEL1"
-            elif message.detail_type == "group":
-                level = "LEVEL2"
-            elif message.detail_type == "channel":
-                level = "LEVEL3"
-            id1 = message.user_id
-            id2 = message.group_id or message.channel_id
-            id3 = message.guild_id
-
-            session_key = (bot_id, bot_type, platform, level, id1, id2, id3)
-            # 保存 message id 和 session key 的对应关系
-            message_id_session_key_map[message.id] = session_key
-            if session_key not in bulk_insert_sessions:
-                bulk_insert_sessions[session_key] = {
-                    "bot_id": bot_id,
-                    "bot_type": bot_type,
-                    "platform": platform,
-                    "level": level,
-                    "id1": id1,
-                    "id2": id2,
-                    "id3": id3,
-                }
+        # 每次迁移的数据量为 10000 条
+        migration_limit = 10000
 
         session_key_id_map = {}
-        if bulk_insert_sessions:
-            # 读取已经存在的 session
-            for session_obj in session.scalars(select(SessionModel)).all():
-                session_key = (
-                    session_obj.bot_id,
-                    session_obj.bot_type,
-                    session_obj.platform,
-                    session_obj.level,
-                    session_obj.id1,
-                    session_obj.id2,
-                    session_obj.id3,
+        count = session.query(MessageRecord).count()
+        for i in range(count // migration_limit + 1):
+            statement = (
+                select(
+                    MessageRecord.bot_id,
+                    MessageRecord.bot_type,
+                    MessageRecord.platform,
+                    MessageRecord.detail_type,
+                    MessageRecord.user_id,
+                    MessageRecord.group_id,
+                    MessageRecord.channel_id,
+                    MessageRecord.guild_id,
+                    MessageRecord.id,
                 )
-                session_key_id_map[session_key] = session_obj.id
+                .limit(migration_limit)
+                .offset(i * migration_limit)
+            )
+            messages = session.execute(statement).all()
 
-            # 更新新插入的 session
-            session.execute(
-                insert(SessionModel),
-                [
+            bulk_insert_sessions = {}
+            message_id_session_key_map = {}
+            for message in messages:
+                bot_id = message[0]
+                bot_type = message[1]
+                platform = message[2]
+                level = "LEVEL0"
+                if message[3] == "private":
+                    level = "LEVEL1"
+                elif message[3] == "group":
+                    level = "LEVEL2"
+                elif message[3] == "channel":
+                    level = "LEVEL3"
+                id1 = message[4]
+                id2 = message[5] or message[6]
+                id3 = message[7]
+
+                session_key = (bot_id, bot_type, platform, level, id1, id2, id3)
+                # 保存 message id 和 session key 的对应关系
+                message_id_session_key_map[message[8]] = session_key
+                if session_key not in bulk_insert_sessions:
+                    bulk_insert_sessions[session_key] = {
+                        "bot_id": bot_id,
+                        "bot_type": bot_type,
+                        "platform": platform,
+                        "level": level,
+                        "id1": id1,
+                        "id2": id2,
+                        "id3": id3,
+                    }
+
+            if bulk_insert_sessions:
+                # 读取已经存在的 session
+                for session_obj in session.scalars(select(SessionModel)).all():
+                    session_key = (
+                        session_obj.bot_id,
+                        session_obj.bot_type,
+                        session_obj.platform,
+                        session_obj.level,
+                        session_obj.id1,
+                        session_obj.id2,
+                        session_obj.id3,
+                    )
+                    session_key_id_map[session_key] = session_obj.id
+
+                # 更新新插入的 session
+                new_sessions_dict = [
                     session_dict
                     for key, session_dict in bulk_insert_sessions.items()
                     if key not in session_key_id_map  # 去重
-                ],
-            )
-            for session_obj in session.scalars(select(SessionModel)).all():
-                session_key = (
-                    session_obj.bot_id,
-                    session_obj.bot_type,
-                    session_obj.platform,
-                    session_obj.level,
-                    session_obj.id1,
-                    session_obj.id2,
-                    session_obj.id3,
-                )
-                session_key_id_map[session_key] = session_obj.id
+                ]
+                if new_sessions_dict:
+                    session.execute(insert(SessionModel), new_sessions_dict)
 
-        # 更新 message 的 session id
-        bulk_update_messages = []
-        for message_id, session_key in message_id_session_key_map.items():
-            bulk_update_messages.append(
-                {
-                    "id": message_id,
-                    "session_id": session_key_id_map[session_key],
-                }
-            )
-        if bulk_update_messages:
-            session.execute(update(MessageRecord), bulk_update_messages)
+                for session_obj in session.scalars(select(SessionModel)).all():
+                    session_key = (
+                        session_obj.bot_id,
+                        session_obj.bot_type,
+                        session_obj.platform,
+                        session_obj.level,
+                        session_obj.id1,
+                        session_obj.id2,
+                        session_obj.id3,
+                    )
+                    session_key_id_map[session_key] = session_obj.id
+
+            # 更新 message 的 session id
+            bulk_update_messages = []
+            for message_id, session_key in message_id_session_key_map.items():
+                bulk_update_messages.append(
+                    {
+                        "id": message_id,
+                        "session_id": session_key_id_map[session_key],
+                    }
+                )
+            if bulk_update_messages:
+                session.execute(update(MessageRecord), bulk_update_messages)
         logger.warning("聊天记录数据迁移完成！")
 
     with op.batch_alter_table(

--- a/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
+++ b/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
@@ -5,6 +5,8 @@ Revises: de6827ead8fe
 Create Date: 2023-06-28 14:44:16.544879
 
 """
+import math
+
 import sqlalchemy as sa
 from alembic import op
 from nonebot.log import logger
@@ -44,9 +46,10 @@ def upgrade() -> None:
 
         # 每次迁移的数据量为 10000 条
         migration_limit = 10000
+        last_message_id = -1
         count = session.query(MessageRecord).count()
         logger.info(f"聊天记录数据总数：{count}")
-        for i in range(count // migration_limit + 1):
+        for i in range(math.ceil(count / migration_limit)):
             statement = (
                 select(
                     MessageRecord.bot_id,
@@ -60,10 +63,11 @@ def upgrade() -> None:
                     MessageRecord.id,
                 )
                 .order_by(MessageRecord.id)
+                .where(MessageRecord.id > last_message_id)
                 .limit(migration_limit)
-                .offset(i * migration_limit)
             )
             messages = session.execute(statement).all()
+            last_message_id = messages[-1][8]
 
             bulk_insert_sessions = {}
             message_id_session_key_map = {}
@@ -85,7 +89,10 @@ def upgrade() -> None:
                 session_key = (bot_id, bot_type, platform, level, id1, id2, id3)
                 # 保存 message id 和 session key 的对应关系
                 message_id_session_key_map[message[8]] = session_key
-                if session_key not in session_key_id_map and session_key not in bulk_insert_sessions: # 去重
+                if (
+                    session_key not in session_key_id_map
+                    and session_key not in bulk_insert_sessions
+                ):  # 去重
                     bulk_insert_sessions[session_key] = {
                         "bot_id": bot_id,
                         "bot_type": bot_type,
@@ -98,7 +105,9 @@ def upgrade() -> None:
 
             if bulk_insert_sessions:
                 # 插入新的 session
-                session.execute(insert(SessionModel), list(bulk_insert_sessions.values()))
+                session.execute(
+                    insert(SessionModel), list(bulk_insert_sessions.values())
+                )
 
                 # 更新已经存在的 session
                 for session_obj in session.scalars(select(SessionModel)).all():

--- a/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
+++ b/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
@@ -59,6 +59,7 @@ def upgrade() -> None:
                     MessageRecord.guild_id,
                     MessageRecord.id,
                 )
+                .order_by(MessageRecord.id)
                 .limit(migration_limit)
                 .offset(i * migration_limit)
             )

--- a/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
+++ b/nonebot_plugin_chatrecorder/migrations/902a51ac4032_add_session.py
@@ -33,6 +33,7 @@ def upgrade() -> None:
 
         session_key_id_map = {}
         count = session.query(MessageRecord).count()
+        logger.info(f"聊天记录数据总数：{count}")
         for i in range(count // migration_limit + 1):
             statement = (
                 select(
@@ -128,6 +129,7 @@ def upgrade() -> None:
                 )
             if bulk_update_messages:
                 session.execute(update(MessageRecord), bulk_update_messages)
+            logger.info(f"已迁移 {i * migration_limit + len(messages)}/{count}")
         logger.warning("聊天记录数据迁移完成！")
 
     with op.batch_alter_table(

